### ISSUE TITLE
can pip-compile handle multiple extras of the same package properly

### ISSUE THREAD

**Issue by zyxue** [jazzband/pip-tools#1511]
#### Environment Versions

1. OS Type: Mac OS or Linux
1. Python version: 3.8.2
1. pip version: 21.3
1. pip-tools version: 6.4.1.dev1+geb9162f

#### Steps to replicate

1. put the following two top-level deps in `requirements.in`
```
cat requirements.in 
ray[default]==1.4.0
tune-sklearn==0.4.0
```
2. `pip-compile requirements.in --rebuild`
3. the output is wrong

My debugging leads to the following:

[`tune-sklearn`](https://github.com/ray-project/tune-sklearn/) depends on [`ray[tune]`](https://github.com/ray-project/ray) as specified in its [setup.py](https://github.com/ray-project/tune-sklearn/blob/master/setup.py#L29), then pip-compile [groups constraints](https://github.com/jazzband/pip-tools/blob/eb9162fdbd6c289ea0ea41699c46fddb3715b242/piptools/resolver.py#L81) into `ray[default,tune]`, which leads to zero dependencies

```
jq .dependencies.ray ~/Library/Caches/pip-tools/depcache-cp3.8.json
{
  "1.4.0[default,tune]": [],
  "1.4.0[default]": [
    "aiohttp",
    "aiohttp-cors",
    "aioredis",
    "click>=7.0",
    "colorama",
    "colorful; extra == \"default\"",
    "filelock",
    "gpustat",
    "grpcio>=1.28.1",
    "jsonschema",
    "msgpack<2.0.0,>=1.0.0",
    "numpy>=1.16",
    "opencensus",
    "prometheus-client>=0.7.1",
    "protobuf>=3.15.3",
    "py-spy>=0.2.0",
    "pydantic>=1.8",
    "pyyaml",
    "redis>=3.5.0",
    "requests"
  ]
}
```

#### Expected result

`"1.4.0[default,tune]"` should have the deps of both `default` and `tune`.

#### Actual result

`"1.4.0[default,tune]"` has zero deps.

**Comment by atugushev** [comment]
Hello @zyxue ,

Thanks for the report. That's because `pip-tools` doesn't yet have a proper dependency resolver. See #693 and the other [related issues](https://github.com/jazzband/pip-tools/issues?q=is%3Aissue+is%3Aopen+label%3Aresolver) for details.

**Comment by zyxue** [comment]
>pip-tools internally uses pip's legacy resolver which is not a complete resolver, see pypa/pip#988.
doesn't pip have a new solver now since pip-20.3 (https://pip.pypa.io/en/stable/user_guide/#changes-to-the-pip-dependency-resolver-in-20-3-2020), is pip-tools planning to use that instead?

Thanks, I'll try to add tests to the PR.

**Comment by atugushev** [comment]
#1539 adds support for the new resolver. Could you give it a try?

**Comment by zyxue** [comment]
tried, still the same, it doesn't fix the example I have at the top

**Comment by di** [comment]
This may be due to https://github.com/pypa/pip/issues/9644.

**Comment by richafrank** [comment]
Hi @zyxue , would you check again on master, now that #1519 has merged and covers this (instead of #1512) ?

**Comment by zyxue** [comment]
@richafrank , tried

* rebase onto https://github.com/jazzband/pip-tools/commit/e2d0fae21769d8ce02f895594aba2407acf9f27e
* run `pip-compile requirements.in --rebuild --annotation-style=line`
* output:

```
jq .dependencies.ray ~/Library/Caches/pip-tools/depcache-cp3.8.json

{
  "1.4.0[default,tune]": [
    "aiohttp",
    "aiohttp-cors",
    "aioredis",
    "click>=7.0",
    "colorama",
    "colorful; extra == \"default\"",
    "filelock",
    "gpustat",
    "grpcio>=1.28.1",
    "jsonschema",
    "msgpack<2.0.0,>=1.0.0",
    "numpy>=1.16",
    "opencensus",
    "pandas; extra == \"tune\"",
    "prometheus-client>=0.7.1",
    "protobuf>=3.15.3",
    "py-spy>=0.2.0",
    "pydantic>=1.8",
    "pyyaml",
    "redis>=3.5.0",
    "requests",
    "tabulate; extra == \"tune\"",
    "tensorboardX; extra == \"tune\""
  ],
  "1.4.0[default]": [
    "aiohttp",
    "aiohttp-cors",
    "aioredis",
    "click>=7.0",
    "colorama",
    "colorful; extra == \"default\"",
    "filelock",
    "gpustat",
    "grpcio>=1.28.1",
    "jsonschema",
    "msgpack<2.0.0,>=1.0.0",
    "numpy>=1.16",
    "opencensus",
    "prometheus-client>=0.7.1",
    "protobuf>=3.15.3",
    "py-spy>=0.2.0",
    "pydantic>=1.8",
    "pyyaml",
    "redis>=3.5.0",
    "requests"
  ]
}
```

so it appears to be working now.
